### DISCUSSION TITLE
Why are tests marked with `#[serial]`?

### DISCUSSION THREAD

**Discussion by giovanniberti** [loco-rs/loco#766] [Q&A] [answered]
Tests by default are marked with `#[serial]`, why is that so?

**Comment by kaplanelad** [comment, marked as answer]
Tests marked with `#[serial]` are primarily used to prevent interference between tests, especially when they involve shared resources like a database.

For example, when running a test that interacts with a database (such as a controller that creates a user), the following steps might occur:

* Clear the database.
* Seed the database with necessary data.
* Make the request to create a user.
* Validate the response.
* Verify the record in the database.

If these steps were run in parallel with other tests, it could lead to conflicts, such as data being overwritten or tests failing unpredictably due to race conditions. By marking the test with `#[serial]`, we ensure that it runs in isolation, preventing one test from affecting another when working with shared resources like a database.

**Reply by giovanniberti** [reply]
Could it be possible to configure `testing::boot_test` to create a separate DB for each test? (or even a different temporary SQLite file)
This way we could run tests that rely on a live DB in parallel, right? :)

**Reply by kaplanelad** [reply]
This setup can works with SQLite using a randomly generated database name. However, in a real-world scenario, you'll want to run the test with the DB engine, which is typically not SQLite :).
